Peerset: reuse Slots data structure for discovered peers

debris opened the pull request:
changes:
- `in_slots`, `out_slots` and `discovered` now utilize the same data structure
- discovered nodes list is now limited to `1000` entries
- `clear_common_slots` is now faster
- `add_peer` is now faster in a case where we need to swap common node with a reserved node
- `Slots` implements custom `fmt::Debug` which is more readable

rphmeier commented on core/peerset/src/slots.rs:
```diff
@@ -104,45 +122,88 @@ impl Slots {
+			.map(|(peer_id, _)| (peer_id, SlotType::Common))
+	}
+
+	pub fn clear_common_slots(&mut self) -> impl Iterator<Item = PeerId> {
```
docs on public items?

rphmeier commented on core/peerset/src/slots.rs:
```diff
@@ -62,39 +83,36 @@ impl Slots {
-			if *st == SlotType::Common && slot_type == SlotType::Reserved {
-				*st = SlotType::Reserved;
+		if self.reserved.contains_key(&peer_id) {
+			return SlotState::AlreadyConnected(peer_id);
```
i notice that it doesn't do a corresponding  "downgrade" of the slot type, which means that there is no invariant at the end of the function that `peer_id` is in a slot of `slot_type`. Can it have any fallout? I would lean towards documenting those behaviors

rphmeier commented on core/peerset/src/slots.rs:
```diff
@@ -104,45 +122,88 @@ impl Slots {
+	/// Pops the oldest peer from the list.
+	pub fn pop_peer(&mut self, reserved_only: bool) -> Option<(PeerId, SlotType)> {
+		if let Some((peer_id, _)) = self.reserved.pop_front() {
+			return Some((peer_id, SlotType::Reserved));
```
It seems kind of strange to pop the reserved peers first. wouldn't we rather cycle out the common ones earlier?

debris replied:
I see where ur confusion comes from. `pop_peer` is only used for `discovered` list. We pop a peer that is later inserted into `out_slots` or `in_slots`

debris commented:
Thanks, I will add/update the docs

tomaka commented:
What's wrong with copy-pasting the code of `Slots` in a different struct?
You seem to be using some methods only in the context of connected nodes, and `pop_peer` only in the context of discovered. Half of the documentation of `Slots` mentions "we are connected to this peer", which is clearly wrong for discovered nodes.

debris commented:
> What's wrong with copy-pasting the code of `Slots` in a different struct?

This is a stupid question. Why do we write functions when we can just copy-paste the code everywhere?

I realized that 6/7 methods in old `Discovered` and exactly the same as in `Slots`. 3/3 properties are the same. There is no point in copying the code.

@rphmeier I updated the docs
@tomaka I also removed references to a `connection` from `Slots` docs in every place where it was misleading

tomaka commented:
In my head, you're going to have to add more fields in the future, such as "how long have we been connected to this node" or "which parachain this slot should be reserved for", and have to split out `Discovered` again.

debris commented:
But we don't have those things now. And it would be premature to keep this code separate because of these reasons

debris commented:
@rphmeier @tomaka can we somehow more forward with this pr?
